utils/swift-xcodegen/Sources/SwiftXcodeGen/Ninja/NinjaBuildFile.swift

```diff
@@ -11,88 +11,168 @@
+
+  struct Bindings: Hashable {
+    let values: [String: String]
+
+    init(storage: [String : String]) {
+      self.values = storage
+    }
+
+    subscript(key: String) -> String? {
+      values[key]
+    }
+  }
+
+  struct Rule {
+    let name: String
+    var bindings: Bindings
+
+    init(name: String, bindings: [String: String]) {
+      self.name = name
+      self.bindings = Bindings(storage: bindings)
+    }
+  }
+
+  struct BuildEdge: Hashable {
+    let ruleName: String
     let inputs: [String]
     let outputs: [String]
     let dependencies: [String]
+    var bindings: Bindings
 
-    let attributes: [Attribute.Key: Attribute]
-    private(set) var isPhony = false
+    var isPhony: Bool {
+      ruleName == "phony"
+    }
 
     init(
+      ruleName: String,
       inputs: [String], outputs: [String], dependencies: [String],
-      attributes: [Attribute.Key : Attribute]
+      bindings: [String: String]
     ) {
+      self.ruleName = ruleName
       self.inputs = inputs
       self.outputs = outputs
       self.dependencies = dependencies
-      self.attributes = attributes
+      self.bindings = Bindings(storage: bindings)
     }
 
     static func phony(for outputs: [String], inputs: [String]) -> Self {
-      var rule = Self(
-        inputs: inputs, outputs: outputs, dependencies: [], attributes: [:]
+      return Self(
+        ruleName: "phony", inputs: inputs, outputs: outputs, dependencies: [], bindings: [:]
       )
-      rule.isPhony = true
-      return rule
     }
   }
 }
 
+
+fileprivate enum NinjaCommandLineError: Error {
+  case unknownRule(String)
+  case missingCommandBinding
+}
+
 extension NinjaBuildFile {
-  struct Attribute: Hashable {
-    var key: Key
-    var value: String
+
+  func commandLine(for edge: BuildEdge) throws -> String {
+    guard  let rule = self.rules[edge.ruleName] else {
+      throw NinjaCommandLineError.unknownRule(edge.ruleName)
+    }
+
+    // Helper to get a substitution value for ${key}.
+    // Note that we don't declare built-in substitions (e.g. $in, $out) for now.
+    func value(for key: String) -> String? {
+      edge.bindings[key] ?? rule.bindings[key] ?? self.bindings[key]
+    }
+
+    func eval(string: String) -> String {
+      var result = ""
+      var rest = string[...]
+      while let dollarIdx = rest.firstIndex(of: "$") {
+        // Eat string before '$'
+        result += rest[rest.startIndex..<dollarIdx]
+        rest = rest[rest.index(after: dollarIdx)...]
+
+        // Parse the key name.
+        let key: String
+        if rest.first == "{" {
+          guard let rightBraceIdx = rest.firstIndex(of: "}") else {
+            result += string[dollarIdx...]
+            return result
+          }
+          key = String(rest[rest.index(after: rest.startIndex)..<rightBraceIdx])
+          rest = rest[rest.index(after: rightBraceIdx)...]
+        } else {
+          let keyEnd = rest.firstIndex(where: { !$0.isLetter && $0 != "_" }) ?? rest.endIndex
```

Comment:
I think numbers are also okay here:

https://github.com/ninja-build/ninja/blob/6b5d7a72975c045ec2bfbb8c6cff6ba359c37289/src/lexer.in.cc#L133